sql/src/main/java/io/crate/execution/engine/window/WindowBatchIterator.java

```diff
@@ -182,16 +219,44 @@ private void computeAndFillStandaloneOutgoingCellsFor(Row sourceRow) {
         }
     }
 
-    private void collectPeers() {
-        Iterable<Row> rows = peerRows.stream().collect(collector);
-        peerRows.clear();
-        peersCollected = true;
-        Row collectedRow = rows.iterator().next();
-        collectorNumberOfColumns = collectedRow.numColumns();
-        for (int i = 0; i < collectorNumberOfColumns; i++) {
-            outgoingCells[i] = collectedRow.get(i);
+    private void executeWindowFunctions() {
+        WindowFrame currentFrame = new WindowFrame(
+            windowRowPosition,
+            sourceRowsConsumed,
+            InMemoryBatchIterator.of(windowForCurrentRow, SentinelRow.SENTINEL)
+        );
+
+        Object[][] cellsForCurrentFrame = executeFunctionsAndUnnestResults(currentFrame);
+        for (Object[] outgoingCells : cellsForCurrentFrame) {
+            resultsForCurrentFrame.push(outgoingCells);
+        }
+
+        windowForCurrentRow.clear();
+    }
+
+    private Object[][] executeFunctionsAndUnnestResults(WindowFrame frame) {
+        int frameSize = windowForCurrentRow.size();
+        Object[][] cellsForCurrentFrame = new Object[frameSize][windowFuncColumnsCount];
+
+        int filledNumColumns = 0;
+        for (WindowFunction function : functions) {
+            Iterator<Row> resultsIterator = function.execute(frame).iterator();
+            // some functions will return only one result (eg. aggregates) but when used in conjunction with functions
```

Comment:
We should probably have some documentation on the `WindowFunction` interface to clarify how `execute` is expected to behave.

I'd have assumed that `execute` always creates a single row but that it is called once per source row (with optional access to the rows in the entire window). This would probably remove the need for the unnesting entirely. Though not sure what other effects it would have.